// @web3modal/core-react-native/lib/commonjs/controllers/OptionsController.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.OptionsController = void 0;
var _valtio = require("valtio");
// -- Types --------------------------------------------- //

// -- State --------------------------------------------- //
const state = (0, _valtio.proxy)({
  projectId: ''
});

// -- Controller ---------------------------------------- //
const OptionsController = exports.OptionsController = {
  state,
  setClipboardClient(client) {
    state._clipboardClient = (0, _valtio.ref)(client);
  },
  setProjectId(projectId) {
    state.projectId = projectId;
  },
  setIncludeWalletIds(includeWalletIds) {
    state.includeWalletIds = includeWalletIds;
  },
  setExcludeWalletIds(excludeWalletIds) {
    state.excludeWalletIds = excludeWalletIds;
  },
  setFeaturedWalletIds(featuredWalletIds) {
    state.featuredWalletIds = featuredWalletIds;
  },
  setTokens(tokens) {
    state.tokens = tokens;
  },
  isClipboardAvailable() {
    return !!state._clipboardClient;
  },
  copyToClipboard(value) {
    const client = state._clipboardClient;
    if (client) {
      client.setString(value);
    }
  }
};
