// @web3modal/scaffold-react-native/lib/module/partials/w3m-header/index.js
import { useSnapshot } from 'valtio';
import { RouterController, ModalController } from '@web3modal/core-react-native';
import { IconLink, Text, FlexView, Separator } from '@web3modal/ui-react-native';
export function Header() {
  const {
    view,
    history
  } = useSnapshot(RouterController.state);
  const headings = () => {
    const name = RouterController.state.data?.wallet?.name;
    const networkName = RouterController.state.data?.network?.name;
    return {
      Connect: 'Connect wallet',
      Account: undefined,
      ConnectingWalletConnect: name ?? 'WalletConnect',
      Networks: 'Select network',
      SwitchNetwork: networkName ?? 'Switch network',
      AllWallets: 'All wallets',
      WhatIsANetwork: 'What is a network?',
      WhatIsAWallet: 'What is a wallet?',
      GetWallet: 'Get a wallet'
    };
  };
  const header = headings()[view];
  const dynamicButtonTemplate = () => {
    const showBack = history.length > 1;
    return showBack ? /*#__PURE__*/React.createElement(IconLink, {
      icon: "chevronLeft",
      size: "md",
      onPress: RouterController.goBack
    }) : /*#__PURE__*/React.createElement(IconLink, {
      icon: "helpCircle",
      size: "md",
      onPress: () => RouterController.push('WhatIsAWallet')
    });
  };
  if (!header) return null;
  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(FlexView, {
    justifyContent: "space-between",
    flexDirection: "row",
    alignItems: "center",
    padding: ['l', 'xl', 'l', 'xl']
  }, dynamicButtonTemplate(), /*#__PURE__*/React.createElement(Text, {
    variant: "paragraph-600",
    numberOfLines: 1
  }, header), /*#__PURE__*/React.createElement(IconLink, {
    icon: "close",
    size: "md",
    onPress: ModalController.close
  })), /*#__PURE__*/React.createElement(Separator, null));
}
