// @web3modal/ui-react-native/lib/commonjs/composites/wui-list-wallet/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ListWallet = ListWallet;
var _reactNative = require("react-native");
var _wuiText = require("../../components/wui-text");
var _useAnimatedValue = _interopRequireDefault(require("../../hooks/useAnimatedValue"));
var _useTheme = require("../../hooks/useTheme");
var _wuiTag = require("../wui-tag");
var _wuiWalletImage = require("../wui-wallet-image");
var _wuiIcon = require("../../components/wui-icon");
var _styles = _interopRequireDefault(require("./styles"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
const AnimatedPressable = _reactNative.Animated.createAnimatedComponent(_reactNative.Pressable);
function ListWallet(_ref) {
  let {
    name,
    onPress,
    imageSrc,
    imageHeaders,
    showAllWallets,
    tagLabel,
    tagVariant,
    icon,
    disabled,
    style
  } = _ref;
  const Theme = (0, _useTheme.useTheme)();
  const {
    animatedValue,
    setStartValue,
    setEndValue
  } = (0, _useAnimatedValue.default)(Theme['gray-glass-002'], Theme['gray-glass-010']);
  function imageTemplate() {
    return /*#__PURE__*/React.createElement(_wuiWalletImage.WalletImage, {
      style: [_styles.default.image, disabled && _styles.default.imageDisabled],
      imageSrc: imageSrc,
      imageHeaders: imageHeaders,
      showAllWallets: showAllWallets,
      size: "sm"
    });
  }
  function iconTemplate() {
    if (tagLabel && tagVariant) {
      return /*#__PURE__*/React.createElement(_wuiTag.Tag, {
        variant: tagVariant,
        disabled: disabled,
        style: _styles.default.rightIcon
      }, tagLabel);
    }
    if (icon) {
      return /*#__PURE__*/React.createElement(_wuiIcon.Icon, {
        name: icon,
        color: disabled ? 'fg-300' : 'fg-200',
        size: "sm",
        style: _styles.default.rightIcon
      });
    }
    return null;
  }
  return /*#__PURE__*/React.createElement(AnimatedPressable, {
    style: [_styles.default.container, {
      backgroundColor: disabled ? Theme['gray-glass-010'] : animatedValue
    }, style],
    disabled: disabled,
    onPress: onPress,
    onPressIn: setEndValue,
    onPressOut: setStartValue
  }, /*#__PURE__*/React.createElement(_reactNative.View, {
    style: _styles.default.leftSide
  }, imageTemplate(), /*#__PURE__*/React.createElement(_wuiText.Text, {
    variant: "paragraph-500",
    style: _styles.default.name,
    color: disabled ? 'fg-300' : 'fg-100'
  }, name)), iconTemplate());
}
