// @web3modal/wagmi-react-native/lib/module/index.js
import '@walletconnect/react-native-compat';
export { W3mAccountButton, W3mButton, W3mConnectButton, Web3Modal } from '@web3modal/scaffold-react-native';
export { defaultWagmiConfig } from './utils/defaultWagmiConfig';
import { useEffect, useState } from 'react';
import { Web3Modal } from './client';
import { VERSION } from './utils/constants';

// -- Types -------------------------------------------------------------------

// -- Setup -------------------------------------------------------------------
let modal;
export function createWeb3Modal(options) {
  if (!modal) {
    modal = new Web3Modal({
      ...options,
      _sdkVersion: `react-native-wagmi-${VERSION}`
    });
  }
  return modal;
}

// -- Hooks -------------------------------------------------------------------
export function useWeb3Modal() {
  if (!modal) {
    throw new Error('Please call "createWeb3Modal" before using "useWeb3Modal" hook');
  }
  async function open(options) {
    await modal?.open(options);
  }
  async function close() {
    await modal?.close();
  }
  return {
    open,
    close
  };
}
export function useWeb3ModalState() {
  if (!modal) {
    throw new Error('Please call "createWeb3Modal" before using "useWeb3ModalState" hook');
  }
  const [state, setState] = useState(modal.getState());
  useEffect(() => {
    const unsubscribe = modal?.subscribeState(newState => {
      if (newState) setState({
        ...newState
      });
    });
    return () => {
      unsubscribe?.();
    };
  }, []);
  return state;
}
